File: framework/src/main/java/ra/hul/framework/pages/LoginPage.java
package ra.hul.framework.pages;


import org.openqa.selenium.By;
import ra.hul.framework.config.ConfigManager;

/**
 * Page Object for the Login page.
 * <p>
 * Rules for Page Objects:
 * 1. Each page = 1 class
 * 2. Locators are PRIVATE (encapsulation)
 * 3. Methods return the NEXT page object (fluent navigation)
 * 4. NO assertions in page objects (that's the test's job)
 * 5. NO WebDriver.get() calls — navigation is explicit
 */
public class LoginPage extends BasePage {

    // ---- Locators (private — encapsulation) ----
    private final By usernameInput = By.id("username");
    private final By passwordInput = By.id("password");
    private final By loginButton = By.cssSelector("button[type='submit']");
    private final By flashMessage = By.id("flash");

    // ---- Actions (public — user behavior) ----

    public LoginPage open() {
        navigateTo(ConfigManager.get("base.url") + "/login");
        return this;
    }

    public LoginPage enterUsername(String username) {
        type(usernameInput, username);
        return this;
    }

    public LoginPage enterPassword(String password) {
        type(passwordInput, password);
        return this;
    }

    public void clickLogin() {
        click(loginButton);
    }

    public String getFlashMessage() {
        return getText(flashMessage).trim();
    }

    // ---- Compound action ----
    public void loginAs(String username, String password) {
        enterUsername(username);
        enterPassword(password);
        clickLogin();
    }

    @Override
    public boolean isLoaded() {
        return isDisplayed(loginButton);
    }
}

File: playwright/pageFactory/ui/LoginPage.ts
import type { Locator, Page } from '@playwright/test';
import { BasePage } from './BasePage';

export class LoginPage extends BasePage {
	readonly heading: Locator;
	readonly subHeading: Locator;
	readonly usernameInput: Locator;
	readonly passwordInput: Locator;
	readonly loginButton: Locator;
	readonly flashMessage: Locator;

	constructor(page: Page) {
		super(page);

		this.heading = page.locator('h2');
		this.subHeading = page.locator('h4');
		this.usernameInput = page.locator('#username');
		this.passwordInput = page.locator('#password');
		this.loginButton = page.locator("button[type='submit']");
		this.flashMessage = page.locator('#flash');
	}

	async goto(): Promise<void> {
		await this.page.goto('/login');
	}

	async login(username: string, password: string): Promise<void> {
		await this.usernameInput.fill(username);
		await this.passwordInput.fill(password);
		await this.loginButton.click();
	}
}
